src/vic3save/src/savefile.rs

```diff
@@ -0,0 +1,225 @@
+use crate::stats::Vic3CountryStats;
+use crate::Vic3Date;
+use serde::de;
+use serde::de::Unexpected;
+use serde::{Deserialize, Deserializer};
+use std::collections::HashMap;
+use std::fmt;
+use std::marker::PhantomData;
+
+#[derive(Debug, Deserialize, PartialEq)]
+pub struct MetaData {
+    pub version: String,
+    pub game_date: Vic3Date,
+}
+
+#[derive(Debug, Deserialize, PartialEq)]
+pub struct Counters {
+    week: i32,
+    tick: i32,
+}
+
+#[derive(Debug, Deserialize, PartialEq)]
+pub struct Vic3CountryBudget {
+    pub weekly_income: Vec<f32>,
+}
+
+#[derive(Debug, Deserialize, PartialEq)]
+pub struct PopStatistics {
+    pub trend_population: Vic3CountryStats,
+}
+#[derive(Debug, Deserialize, PartialEq)]
+pub struct Vic3Country {
+    pub definition: String,
+    pub government: String,
+    pub budget: Vic3CountryBudget,
+    pub gdp: Vic3CountryStats,
+    pub literacy: Vic3CountryStats,
+    pub prestige: Vic3CountryStats,
+    pub avgsoltrend: Vic3CountryStats,
+    pub pop_statistics: PopStatistics,
+}
+#[derive(Debug, PartialEq)]
```

Comment:
I know this suggestion results in more code, but I think it is worth it so `Maybe` isn't part of the public interface:

```diff
diff --git a/src/vic3save/src/savefile.rs b/src/vic3save/src/savefile.rs
index 03b77c2..c2d628f 100644
--- a/src/vic3save/src/savefile.rs
+++ b/src/vic3save/src/savefile.rs
@@ -2,9 +2,10 @@ use crate::stats::Vic3CountryStats;
 use crate::Vic3Date;
 use serde::de;
 use serde::de::Unexpected;
-use serde::{Deserialize, Deserializer};
+use serde::{de::DeserializeOwned, Deserialize, Deserializer};
 use std::collections::HashMap;
 use std::fmt;
+use std::hash::Hash;
 use std::marker::PhantomData;
 
 #[derive(Debug, Deserialize, PartialEq)]
@@ -39,11 +40,9 @@ pub struct Vic3Country {
     pub avgsoltrend: Vic3CountryStats,
     pub pop_statistics: PopStatistics,
 }
+
 #[derive(Debug, PartialEq)]
-pub enum Maybe<T> {
-    Some(T),
-    None,
-}
+struct Maybe<T>(Option<T>);
 impl<'de, T> de::Deserialize<'de> for Maybe<T>
 where
     T: Deserialize<'de>,
@@ -72,7 +71,7 @@ where
                 E: serde::de::Error,
             {
                 match v {
-                    "none" => Ok(Maybe::None),
+                    "none" => Ok(Maybe(None)),
                     _ => Err(E::invalid_value(Unexpected::Other(v), &self)),
                 }
             }
@@ -81,7 +80,7 @@ where
             where
                 A: de::MapAccess<'de>,
             {
-                T1::deserialize(de::value::MapAccessDeserializer::new(map)).map(Maybe::Some)
+                T1::deserialize(de::value::MapAccessDeserializer::new(map)).map(|x| Maybe(Some(x)))
             }
         }
         deserializer.deserialize_any(MaybeVisitor {
@@ -90,9 +89,52 @@ where
     }
 }
 
+fn maybe_map<'de, D, K, V>(deser: D) -> Result<HashMap<K, Option<V>>, D::Error>
+where
+    D: Deserializer<'de>,
+    K: Deserialize<'de> + Hash + Eq,
+    V: Deserialize<'de>,
+{
+    struct MaybeVisitor<K, V> {
+        marker: PhantomData<HashMap<K, Option<V>>>,
+    }
+
+    impl<'de, K1, V1> de::Visitor<'de> for MaybeVisitor<K1, V1>
+    where
+        K1: Deserialize<'de> + Hash + Eq,
+        V1: Deserialize<'de>,
+    {
+        type Value = HashMap<K1, Option<V1>>;
+
+        fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+            formatter.write_str("A maybe map")
+        }
+
+        fn visit_map<A>(self, mut access: A) -> Result<Self::Value, A::Error>
+        where
+            A: de::MapAccess<'de>,
+        {
+            let mut map = HashMap::with_capacity(access.size_hint().unwrap_or(0));
+            while let Some((key, value)) = access.next_entry::<K1, Maybe<V1>>()? {
+                map.insert(key, value.0);
+            }
+
+            Ok(map)
+        }
+    }
+
+    deser.deserialize_map(MaybeVisitor {
+        marker: PhantomData,
+    })
+}
+
 #[derive(Debug, Deserialize, PartialEq)]
-pub struct Vic3Manager<Of> {
-    pub database: HashMap<u32, Of>,
+pub struct Vic3Manager<Of>
+where
+    Of: DeserializeOwned,
+{
+    #[serde(deserialize_with = "maybe_map")]
+    pub database: HashMap<u32, Option<Of>>,
 }
 
 #[derive(Debug, Deserialize, PartialEq)]
@@ -105,7 +147,7 @@ pub struct Player {
 pub struct Vic3Save {
     pub meta_data: MetaData,
     pub counters: Counters,
-    pub country_manager: Vic3Manager<Maybe<Vic3Country>>,
+    pub country_manager: Vic3Manager<Vic3Country>,
     pub previous_played: Vec<Player>,
 }
 
@@ -116,7 +158,7 @@ impl Vic3Save {
             .database
             .iter()
             .find_map(|(_, &ref country)| match country {
-                Maybe::Some(
+                Some(
                     country @ Vic3Country {
                         definition: def, ..
                     },
@@ -128,7 +170,7 @@ impl Vic3Save {
     pub fn get_last_played_country<'a>(&'a self) -> &'a Vic3Country {
         let mngr = &self.country_manager;
         let last_played_id = &self.previous_played[0].idtype;
-        let Maybe::Some(ref country) = mngr.database[last_played_id] else {
+        let Some(ref country) = mngr.database[last_played_id] else {
             panic!("played country not found")
         };
         return country;
@@ -137,7 +179,7 @@ impl Vic3Save {
 
 #[cfg(test)]
 mod tests {
-    use crate::savefile::{Counters, Maybe, MetaData, Vic3CountryBudget, Vic3Manager};
+    use super::*;
     use jomini::text::de::from_utf8_slice;
     use serde::Deserialize;
     use std::collections::HashMap;
@@ -152,7 +194,7 @@ mod tests {
     pub struct Vic3Save {
         pub meta_data: MetaData,
         pub counters: Counters,
-        pub country_manager: Vic3Manager<Maybe<Vic3Country>>,
+        pub country_manager: Vic3Manager<Vic3Country>,
     }
 
     const SAVE_FILE: &[u8] = br#"
@@ -189,7 +231,7 @@ counters={
     #[test]
     fn test_full_save() {
         let out: Vic3Save = from_utf8_slice(SAVE_FILE).expect("test");
-        let Maybe::Some(ref country) = out.country_manager.database[&16777216] else {
+        let Some(ref country) = out.country_manager.database[&16777216] else {
             panic!()
         };
         assert_eq!(country.definition, "GER");
@@ -208,15 +250,16 @@ budget={
 
         #[derive(Debug, Deserialize, PartialEq)]
         struct TestCountry {
-            country: HashMap<u32, Maybe<Vic3Country>>,
+            #[serde(deserialize_with = "maybe_map")]
+            country: HashMap<u32, Option<Vic3Country>>,
         }
         let out: TestCountry = from_utf8_slice(country_segment).expect("test");
-        let Maybe::Some(ref country) = out.country[&101] else {
+        let Some(ref country) = out.country[&101] else {
             panic!()
         };
         assert_eq!(country.definition, "GER");
         let out1: TestCountry = from_utf8_slice(b"country={101=none}").expect("test");
-        assert_eq!(out1.country[&101], Maybe::None);
+        assert_eq!(out1.country[&101], None);
         assert_eq!(
             true,
             from_utf8_slice::<TestCountry>(b"country={101=None}").is_err()
```

How do you feel about this?

Reply:
Yeah, not exposing the Maybe type is nicer, thanks!